AudioDeviceExample/AudioDevices/ExampleAudioEngineDevice.m

```diff
@@ -0,0 +1,741 @@
+
+@property (nonatomic, strong) AVAudioEngine *engine;
+@property (nonatomic, strong) AVAudioPlayerNode *player;
+@property (nonatomic, strong) AVAudioUnitReverb *reverb;
+
+@end
+
+@implementation ExampleAudioEngineDevice
+
+#pragma mark - Init & Dealloc
+
+- (id)init {
+    if (@available(iOS 11.0, *)) {
+        self = [super init];
+
+        if (self) {
+        }
+
+        return self;
+    } else {
+        self = nil;
+        NSException *exception = [NSException exceptionWithName:@"ExampleAudioEngineDeviceNotSupported"
+                                                         reason:@"ExampleAudioEngineDevice requires iOS 11.0 or greater." userInfo:nil];
+        [exception raise];
+        return self;
+    }
+}
+
+- (void)dealloc {
+    [self unregisterAVAudioSessionObservers];
+}
+
++ (NSString *)description {
+    return @"AVAudioEngine Audio Mixing";
+}
+
+/*
+ * Determine at runtime the maximum slice size used by RemoteIO. Setting the stream format and sample rate doesn't
+ * appear to impact the maximum size so we prefer to read this value once at initialization time.
+ */
++ (void)initialize {
+    AudioComponentDescription audioUnitDescription = [self audioUnitDescription];
+    AudioComponent audioComponent = AudioComponentFindNext(NULL, &audioUnitDescription);
+    AudioUnit audioUnit;
+    OSStatus status = AudioComponentInstanceNew(audioComponent, &audioUnit);
+    if (status != 0) {
+        NSLog(@"Could not find RemoteIO AudioComponent instance!");
+        return;
+    }
+
+    UInt32 framesPerSlice = 0;
+    UInt32 propertySize = sizeof(framesPerSlice);
+    status = AudioUnitGetProperty(audioUnit, kAudioUnitProperty_MaximumFramesPerSlice,
+                                  kAudioUnitScope_Global, kOutputBus,
+                                  &framesPerSlice, &propertySize);
+    if (status != 0) {
+        NSLog(@"Could not read RemoteIO AudioComponent instance!");
+        AudioComponentInstanceDispose(audioUnit);
+        return;
+    }
+
+    NSLog(@"This device uses a maximum slice size of %d frames.", (unsigned int)framesPerSlice);
+    kMaximumFramesPerBuffer = (size_t)framesPerSlice;
+    AudioComponentInstanceDispose(audioUnit);
+}
+
+#pragma mark - AudioEngine
+
+- (void)setupAudioEngine {
+    NSAssert(_engine == nil, @"AVAudioEngine is already configured");
+
+    _engine = [AVAudioEngine new];
+    [_engine stop];
+
+    NSError *error = nil;
+    const AudioStreamBasicDescription asbd = [self.renderingFormat streamDescription];
+    AVAudioFormat *format = [[AVAudioFormat alloc] initWithStreamDescription:&asbd];
+    if (@available(iOS 11.0, *)) {
+        BOOL success = [_engine enableManualRenderingMode:AVAudioEngineManualRenderingModeRealtime
+                                                   format:format
+                                        maximumFrameCount:(uint32_t)kMaximumFramesPerBuffer
+                                                    error:&error];
```

Comment:
Should we ask for exactly 10 msec of audio here? It seems like we're adding extra delay by using the maximum, but I don't know all the details of AVAudioEngine.